resilience4j-core/src/main/java/io/github/resilience4j/core/metrics/SlidingTimeWindowMetrics.java

```diff
@@ -0,0 +1,134 @@
+/*
+ *
+ *  Copyright 2019 Robert Winkler and Bohdan Storozhuk
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.core.metrics;
+
+
+import java.time.Clock;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A {@link Metrics} implementation is backed by a sliding time window that aggregates only the calls made
+ * in the last {@code N} seconds.
+ *
+ * The sliding time window is implemented with a circular array of {@code N} partial aggregations (buckets).
+ * If the time window size is 10 seconds, the circular array has always 10 partial aggregations (buckets).
+ * Every bucket aggregates the outcome of all calls which happen in a certain epoch second. (Partial aggregation)
+ * The head bucket of the circular array stores the call outcomes of the current epoch second.
+ * The other partial aggregations store the call outcomes of the previous {@code N-1} epoch seconds.
+ *
+ * The sliding window does not store call outcomes (tuples) individually, but incrementally updates partial aggregations (bucket) and a total aggregation.
+ * The total total aggregation is updated incrementally when a new call outcome is recorded. When the oldest bucket is evicted, the partial total aggregation of that bucket
+ * is subtracted from the total totalAggregation and the bucket is reset. (Subtract-on-Evict)
+ *
+ * The time to retrieve a Snapshot is constant 0(1), since the Snapshot is pre-aggregated and is independent of the time window size.
+ * The space requirement (memory consumption) of this implementation should be nearly constant O(n), since the call outcomes (tuples) are not stored individually.
+ * Only {@code N} partial aggregations and 1 total total aggregation are created.
+ */
+public class SlidingTimeWindowMetrics implements Metrics {
+
```

Comment:
I feel we should separate the sliding window logic and storage into separate class aside from the metrics , so anyone want to use outside the metrics Interface , he can do it , i am thinking about the adaptive bulkhead case here , what do u think ?